File: routes/verify.js
// routes/verify.js
import express from "express";
import { db, admin } from "../firebase.js";

const router = express.Router();

/**
 * POST /auth/verify
 * Frontend sends Authorization: Bearer <Firebase ID Token>
 */
router.post("/", async (req, res) => {
  try {
    const authHeader = req.headers.authorization;
    console.log("🔎 Received authHeader:", authHeader);

    if (!authHeader || !authHeader.startsWith("Bearer ")) {
      return res.status(401).json({ message: "Missing or invalid token" });
    }

    const idToken = authHeader.split(" ")[1];
    console.log("🔑 Token received:", idToken.substring(0, 20) + "...");

    // Verify Firebase ID token using admin SDK
    const decodedToken = await admin.auth().verifyIdToken(idToken);
    console.log("✅ Decoded Token:", decodedToken);

    const uid = decodedToken.uid;

    // Fetch user document from Firestore using admin credentials
    const userRef = db.collection("users").doc(uid);
    const userDoc = await userRef.get();

    if (!userDoc.exists) {
      console.warn(`⚠️ User not found in Firestore: UID=${uid}`);
      return res.status(404).json({ message: "User not found in database." });
    }

    const userData = userDoc.data();
    const role = userData.role || "unknown";

    // Optional: validate role sent from frontend
    const frontendRole = req.body.role?.toLowerCase();
    if (frontendRole && frontendRole !== role.toLowerCase()) {
      console.warn(`⚠️ Role mismatch: frontend=${frontendRole}, backend=${role}`);
      return res.status(403).json({ message: "Role mismatch. Access denied." });
    }

    return res.json({
      success: true,
      uid,
      email: decodedToken.email,
      role,
    });

  } catch (err) {
    console.error("❌ Token verification or Firestore error:", err);

    // Differentiate errors
    if (err.code === "auth/id-token-expired") {
      return res.status(401).json({ message: "Token expired" });
    }
    if (err.code === "auth/argument-error") {
      return res.status(400).json({ message: "Invalid token format" });
    }

    return res.status(401).json({ message: "Invalid or expired token" });
  }
});

export default router;

File: route/verify.py
# routes/verify.py
from flask import Blueprint, request, jsonify
from firebase_admin import auth, firestore
from app import db  # Assuming your app.py exports Firestore client as db

verify_bp = Blueprint("verify", __name__)

@verify_bp.route("/", methods=["POST"])
def verify_user():
    try:
        auth_header = request.headers.get("Authorization", "")
        print("🔎 Received authHeader:", auth_header)

        if not auth_header or not auth_header.startswith("Bearer "):
            return jsonify({"message": "Missing or invalid token"}), 401

        id_token = auth_header.split(" ")[1]
        print("🔑 Token received:", id_token[:20] + "...")

        # Verify Firebase ID token using admin SDK
        decoded_token = auth.verify_id_token(id_token)
        print("✅ Decoded Token:", decoded_token)

        uid = decoded_token["uid"]

        # Fetch user document from Firestore
        user_ref = db.collection("users").document(uid)
        user_doc = user_ref.get()

        if not user_doc.exists:
            print(f"⚠️ User not found in Firestore: UID={uid}")
            return jsonify({"message": "User not found in database."}), 404

        user_data = user_doc.to_dict()
        role = user_data.get("role", "unknown")

        # Optional: validate role sent from frontend
        frontend_role = request.json.get("role", "").lower() if request.json else None
        if frontend_role and frontend_role != role.lower():
            print(f"⚠️ Role mismatch: frontend={frontend_role}, backend={role}")
            return jsonify({"message": "Role mismatch. Access denied."}), 403

        return jsonify({
            "success": True,
            "uid": uid,
            "email": decoded_token.get("email"),
            "role": role
        })

    except auth.ExpiredIdTokenError:
        return jsonify({"message": "Token expired"}), 401
    except auth.InvalidIdTokenError:
        return jsonify({"message": "Invalid token format"}), 400
    except Exception as e:
        print("❌ Token verification or Firestore error:", e)
        return jsonify({"message": "Invalid or expired token"}), 401
